From 8cda2a14fd602fa16ace65e06c74c5e48a372c22 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Thu, 26 Nov 2020 21:12:39 +0100
Subject: [PATCH 1/1] Kconfig: allow compiling for debugging

Currently via Kconfig we can choose between size and performance
optimization.

Add an option to compile for debugging which passes -Og to GCC.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 Kconfig                       | 23 ++++++++++++++++++++---
 Makefile                      |  6 +++++-
 configs/Cyrus_P5020_defconfig |  2 +-
 configs/Cyrus_P5040_defconfig |  2 +-
 4 files changed, 27 insertions(+), 6 deletions(-)

diff --git a/Kconfig b/Kconfig
index 6dc20ed25b..3e4d5f979e 100644
--- a/Kconfig
+++ b/Kconfig
@@ -57,15 +57,32 @@ config LOCALVERSION_AUTO
 
 	  which is done within the script "scripts/setlocalversion".)
 
+choice
+	prompt "Optimization target for compiler"
+	default CC_OPTIMIZE_FOR_SIZE
+
 config CC_OPTIMIZE_FOR_SIZE
 	bool "Optimize for size"
-	default y
 	help
-	  Enabling this option will pass "-Os" instead of "-O2" to gcc
-	  resulting in a smaller U-Boot image.
+	  Enabling this option will pass "-Os" to GCC resulting in
+	  a smaller U-Boot image.
 
 	  This option is enabled by default for U-Boot.
 
+config CC_OPTIMIZE_FOR_PERFORMANCE
+	bool "Optimize for performance"
+	help
+	  Enabling this option will pass "-O2" to GCC resulting in
+	  a better performance.
+
+config CC_OPTIMIZE_FOR_DEBUGGING
+	bool "Optimize for debugging"
+	help
+	  Enabling this option will pass "-Og" to GCC resulting in
+	  an easy to debug U-Boot image.
+
+endchoice
+
 config OPTIMIZE_INLINING
 	bool "Allow compiler to uninline functions marked 'inline' in full U-Boot"
 	default n
diff --git a/Makefile b/Makefile
index 17719666fb..850ecf8615 100644
--- a/Makefile
+++ b/Makefile
@@ -673,9 +673,13 @@ endif
 
 ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS	+= -Os
-else
+endif
+ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE
 KBUILD_CFLAGS	+= -O2
 endif
+ifdef CONFIG_CC_OPTIMIZE_FOR_DEBUGGING
+KBUILD_CFLAGS	+= -Og
+endif
 
 KBUILD_CFLAGS += $(call cc-option,-fno-stack-protector)
 KBUILD_CFLAGS += $(call cc-option,-fno-delete-null-pointer-checks)
diff --git a/configs/Cyrus_P5020_defconfig b/configs/Cyrus_P5020_defconfig
index a488ad8fb5..615ff1a563 100644
--- a/configs/Cyrus_P5020_defconfig
+++ b/configs/Cyrus_P5020_defconfig
@@ -4,7 +4,7 @@ CONFIG_ENV_SIZE=0x2000
 CONFIG_ENV_OFFSET=0xCF400
 CONFIG_MPC85xx=y
 CONFIG_TARGET_CYRUS_P5020=y
-# CONFIG_CC_OPTIMIZE_FOR_SIZE is not set
+CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE=y
 CONFIG_FIT=y
 CONFIG_FIT_VERBOSE=y
 CONFIG_OF_BOARD_SETUP=y
diff --git a/configs/Cyrus_P5040_defconfig b/configs/Cyrus_P5040_defconfig
index 476d63bab2..d4b3b58399 100644
--- a/configs/Cyrus_P5040_defconfig
+++ b/configs/Cyrus_P5040_defconfig
@@ -4,7 +4,7 @@ CONFIG_ENV_SIZE=0x2000
 CONFIG_ENV_OFFSET=0xCF400
 CONFIG_MPC85xx=y
 CONFIG_TARGET_CYRUS_P5040=y
-# CONFIG_CC_OPTIMIZE_FOR_SIZE is not set
+CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE=y
 CONFIG_FIT=y
 CONFIG_FIT_VERBOSE=y
 CONFIG_OF_BOARD_SETUP=y
-- 
2.29.2

